### PR TITLE
Allow extra keys in config

### LINKED ISSUE #43: Configuring Auth0 SdkConfiguration arguments
Hi @Nyholm,
The `auth0-php:@8.0.0-BETA1` fail to issue login link without defining `cookieSecret` parameter for `SdkConfiguration` (however defined as optionnal in code and documentation).
Configuration injected in the `SdkConfiguration` $oponfiguration argument is fixed the `ConfigurationProvider` service declaration.
How can we allow to define all available configuration:
  * Take a free array in the bundle configuration and let auth0 do the validation ?
  * Specify every SdkConfiguration constructor argument explicity in the bundle configuration to validate eveything is well configured before pass arguments to auth0 configuration ? 

What do you think ?

### PR DESCRIPTION
**Pull request by Nyholm**
This will fix #43

FYI @amenophis 

-------

It will move the values sent to `SdkConfiguration` to a new `config` keyword. 

```yaml
happyr_auth0:
    httplug_client_service: 'app.test.httplug'
    config:
        domain: 'foo.com'
        clientId: 'client_id'
        clientSecret: 'secret'
        scope:
            - 'foo'
            - 'bar'
    firewall:
        check_route: default_login_check
        failure_path: startpage
        default_target_path: user_login
```

Under the `config` keyword once can add an arbitrary key/value pair. 

It will also remove all container parameters (`auth0.*`). The last thing it does is to make sure "scope" is an array.

### CODE REVIEW

File: src/DependencyInjection/HappyrAuth0Extension.php [hunk 1]
```diff
@@ -50,6 +50,13 @@ public function load(array $configs, ContainerBuilder $container)
 
         $configProviderDefinition = $container->getDefinition(ConfigurationProvider::class);
 
+        // if extra keys
+        if (isset($configs[1]) && [] !== $configs[1]) {
```
Comment (amenophis) [inline src/DependencyInjection/HappyrAuth0Extension.php:54]: Why $configs[1] ?

**Comment by amenophis** [comment]
I see an issue with this implementation, the bundle use keys in snake_kase (client_id), and the sdk seems to use keys in CamelCase (cookieSecret).

It can lead to mixed format which is not very good ...

File: src/DependencyInjection/HappyrAuth0Extension.php (hunk 1 above)
Comment (Nyholm) [reply src/DependencyInjection/HappyrAuth0Extension.php:54]: This is the first time Im using the "extra keys feature". 

So `$configs` is an array of config values. Those are processed in `$this->processConfiguration()` into a `$config` array. 

The `$configs[1]` will always include the "extra keys". (if my debugging is correct)

**Comment by Nyholm** [comment]
Great observation! Hm.. 

Should we convert all snake_case to camelCase for the extra keys?.. Or should we change all our existing config to match the SDK?

File: src/DependencyInjection/HappyrAuth0Extension.php (hunk 1 above)
Comment (amenophis) [reply src/DependencyInjection/HappyrAuth0Extension.php:54]: I don't think it work like that.
If i remember correctly, the $configs array contain all `happyr_auth0` nodes if confirgured in multiples files.
You need to use only the result of `$this->processConfiguration()` which is in chage to aggregate all of them.
The result contains declared keys alongside extrax keys.

**Comment by amenophis** [comment]
In YAML, the habit is to user snake_case, so i think we must use snake_case keys and transform them to CamelCase to inject key/values in the SdkConfiguration

File: src/DependencyInjection/HappyrAuth0Extension.php (hunk 1 above)
Comment (Nyholm) [reply src/DependencyInjection/HappyrAuth0Extension.php:54]: Oh. you might be correct. I'll do some better debugging. 

Thank you

**Comment by amenophis** [comment]
I think it could be better to move all auth0 SdkConfiguration related keys to a subnode named auth0 and handle all keys in the same way ?

**Comment by amenophis** [comment]
I just see your last commit, i was trying something similar.
Just one exception, assign dynamically all arguments like this:

```php
$sdkConfigurationDefinition = $container->getDefinition(SdkConfiguration::class);
foreach ($config['sdk'] ?? [] as $key => $value) {
    $argumentName = '$'.lcfirst(str_replace('_', '', ucwords($key, '_')));
    $sdkConfigurationDefinition->setArgument($argumentName, $value);
}
```

What do you think ?

**Comment by amenophis** [comment]
The previous proposition didn't work, the Symfony container complains parameters are not correctly configured ...
I tried another solution with reflection and it seems to work !

```php
$camelConfig = [];
foreach ($config['sdk'] ?? [] as $key => $value) {
    $camelKey = lcfirst(str_replace('_', '', ucwords($key, '_')));
    $camelConfig[$camelKey] = $value;
}

$sdkConfigurationRefClass = new \ReflectionClass(SdkConfiguration::class);
$constructor = $sdkConfigurationRefClass->getConstructor();

$sdkConfigurationDefinition = $container->getDefinition(SdkConfiguration::class);
foreach ($constructor->getParameters() as $parameter) {
    $sdkConfigurationDefinition->setArgument(
        '$'.$parameter->getName(),
        $camelConfig[$parameter->getName()] ?? $parameter->getDefaultValue()
    );
}
```

Is it OK for you ?

**Comment by Nyholm** [comment]
Sorry. I was in a rush to leave. I didn't finish my last commit. 

The `SdkConfiguration`'s constructor is a bit tricky to read. But it accepts an associative array. So I don't thing we should try to convert from snake to camel etc. This is simpler and it will always work, right?

So currently we have 8 defined "config keys". Im happy to add more. When the SDK adds a new config key, users can use it without the bundle making an update.

**Comment by amenophis** [comment]
The documentation encourage to use the namedArguments, and say the array is just to simplify usage with 7.4 (https://github.com/auth0/auth0-PHP#configuration-options)

The main advantage I see to use the arguments directly is we can rely on the default values provided but constructor arguments default values.

I do not see any advantage to have some harcoded options but not all, so i think we need to provide all or nothing.
Maybe we can go further with reflection to extract documentation generate the TreeBuilder with reflection directly ?

I can made a try on this if you want

**Comment by amenophis** [comment]
I made a first try and can be able to generate something like this in the tree builder from constructor docblock:
image

I'm waiting for your update to finish the implementation

**Comment by amenophis** [comment]
This is the output of symfony debug:dump-reference command:

```bash
$ bin/console config:dump-reference HappyrAuth0Bundle
# Default configuration for "HappyrAuth0Bundle"
happyr_auth0:
    sdk:

        # Required. Auth0 domain for your tenant.
        domain:               null

        # Required. Client ID, found in the Auth0 Application settings.
        clientId:             null

        # Optional. Authentication callback URI, as defined in your Auth0 Application settings.
        redirectUri:          null

        # Optional. Client Secret, found in the Auth0 Application settings.
        clientSecret:         null

        # Optional. One or more API identifiers, found in your Auth0 API settings. The SDK uses the first value for building links. If provided, at least one of these values must match the 'aud' claim to validate an ID Token successfully.
        audience:             null

        # Optional. One or more Organization IDs, found in your Auth0 Organization settings. The SDK uses the first value for building links. If provided, at least one of these values must match the 'org_id' claim to validate an ID Token successfully.
        organization:         null

        # Optional. Defaults to true. Use PKCE (Proof Key of Code Exchange) with Authorization Code Flow requests. See https://auth0.com/docs/flows/call-your-api-using-the-authorization-code-flow-with-pkce
        usePkce:              true

        # Optional. One or more scopes to request for Tokens. See https://auth0.com/docs/scopes
        scope:

            # Defaults:
            - openid
            - profile
            - email

        # Optional. Defaults to 'query.' Where to extract request parameters from, either 'query' for GET or 'form_post' for POST requests.
        responseMode:         query

        # Optional. Defaults to 'code.' Use 'code' for server-side flows and 'token' for application side flow.
        responseType:         code

        # Optional. Defaults to 'RS256'. Algorithm to use for Token verification. Expects either 'RS256' or 'HS256'.
        tokenAlgorithm:       RS256

        # Optional. URI to the JWKS when verifying RS256 tokens.
        tokenJwksUri:         null

        # Optional. The maximum window of time (in seconds) since the 'auth_time' to accept during Token validation.
        tokenMaxAge:          null

        # Optional. Defaults to 60. Leeway (in seconds) to allow during time calculations with Token validation.
        tokenLeeway:          60

        # Optional. A PSR-6 compatible cache adapter for storing JSON Web Key Sets (JWKS).
        tokenCache:           null

        # Optional. How long (in seconds) to keep a JWKS cached.
        tokenCacheTtl:        60

        # Optional. A PSR-18 compatible HTTP client to use for API requests.
        httpClient:           null

        # Optional. A PSR-17 compatible request factory to generate HTTP requests.
        httpRequestFactory:   null

        # Optional. A PSR-17 compatible response factory to generate HTTP responses.
        httpResponseFactory:  null

        # Optional. A PSR-17 compatible stream factory to create request body streams.
        httpStreamFactory:    null

        # Optional. Defaults to true. If true, API requests will include telemetry about the SDK and PHP runtime version to help us improve our services.
        httpTelemetry:        true

        # Optional. Defaults to use cookies. A StoreInterface-compatible class for storing Token state.
        sessionStorage:       null

        # Optional. The secret used to derive an encryption key for the user identity in a session cookie and to sign the transient cookies used by the login callback.
        cookieSecret:         null

        # Optional. Defaults to value of HTTP_HOST server environment information. Cookie domain, for example 'www.example.com', for use with PHP sessions and SDK cookies. To make cookies visible on all subdomains then the domain must be prefixed with a dot like '.example.com'.
        cookieDomain:         null

        # Optional. Defaults to 0. How long, in seconds, before cookies expire. If set to 0 the cookie will expire at the end of the session (when the browser closes).
        cookieExpires:        0

        # Optional. Defaults to '/'. Specifies path on the domain where the cookies will work. Use a single slash ('/') for all paths on the domain.
        cookiePath:           /

        # Optional. Defaults to false. Specifies whether cookies should ONLY be sent over secure connections.
        cookieSecure:         false

        # Optional. Defaults to true. If true, the user data will persist in session storage.
        persistUser:          true

        # Optional. Defaults to true. If true, the Id Token will persist in session storage.
        persistIdToken:       true

        # Optional. Defaults to true. If true, the Access Token will persist in session storage.
        persistAccessToken:   true

        # Optional. Defaults to true. If true, the Refresh Token will persist in session storage.
        persistRefreshToken:  true

        # Optional. Defaults to use cookies. A StoreInterface-compatible class for storing ephemeral state data, such as nonces.
        transientStorage:     null

        # Optional. Defaults to false. If true, query the /userinfo endpoint during an authorization code exchange.
        queryUserInfo:        false

        # Optional. An Access Token to use for Management API calls. If there isn't one specified, the SDK will attempt to get one for you using your $clientSecret.
        managementToken:      null

        # Optional. A PSR-6 compatible cache adapter for storing generated management access tokens.
        managementTokenCache: null

        # Optional. A PSR-14 compatible event listener provider, for interfacing with events triggered by the SDK.
        eventListenerProvider: null
```

I think it is a good way because it auto-document the bundle configuration for the sdk node.
For this version, everything is a scalarNode, maybe we need to manage node type based on reflection type ?

What do you think ?

**Comment by Nyholm** [comment]
I do like the fact that it is auto generated and that we have the default values. 

I think you are in to something. Let's merge this PR (since it is doing some unrelated things are are moving in the correct direction). We should merge a variation of #46 before releasing anything.

**Pull request merged**